Jeremy brouillet/add mmr

jeremy-brouillet opened the pull request:
þ

# Description

This change adds a maximum marginal relevance option for the SimpleVectorStore. By enabling the mmr query mode, the SimpleVectorStore will change from comparing vectors based purely on similarity to adding a diversity factor based on similarity to previous results. Additionally, a notebook was added with a simple example as well as tests.

Fixes # (issue)

No issues.

## Type of Change

Please delete options that are not relevant.

- [x] New feature (non-breaking change which adds functionality)
- [x] This change requires a documentation update

# How Has This Been Tested?

Please describe the tests that you ran to verify your changes. Provide instructions so we can reproduce. Please also list any relevant details for your test configuration

- [x] Added new unit/integration tests
- [x] Added new notebook (that tests end-to-end)
- [x] I stared at the code and made sure it makes sense

# Suggested Checklist:

- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation
- [x] My changes generate no new warnings
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes

logan-markewich commented on llama_index/indices/query/embedding_utils.py:
```diff
@@ -92,3 +92,62 @@ def get_top_k_embeddings_learner(
+
+    for i, emb in enumerate(embeddings):
+        similarity = similarity_fn(query_embedding, emb)
+        similarities.append([similarity, embedding_ids[i], 0])
```
nit: can you use a tuple here instead? `(similarity, embedding_ids[i], 0)`

jeremy-brouillet replied:
The reason I'm using a tuple is that the third entry, the zero, is the diversity score and is updated. Tuples don't support item assignment.

logan-markewich replied:
Right!

logan-markewich commented on llama_index/vector_stores/simple.py:
```diff
@@ -147,6 +150,22 @@ def query(
                 embedding_ids=node_ids,
             )
+        elif query.mode == MMR_MODE:
+            if "mmr_threshold" in kwargs:
```
instead of this if/else, you can just do `mmr_threshold=kwargs.get("mmr_threshold", None)`

jeremy-brouillet replied:
Good point, will add that.

jeremy-brouillet commented on llama_index/indices/query/embedding_utils.py:
```diff
@@ -92,3 +92,62 @@ def get_top_k_embeddings_learner(
     return result_similarities, result_ids
+
+
+def get_top_k_mmr_embeddings(
```
Similarity will go as O(n log(n)), because you're adding elements to the heap, whereas MMR goes as O(n^2 log(n)), as each result could affect the other results and there's a re-sort at every step. It's an iterative algorithm. I agree that performance could suffer at scale.

What is best practice for llama index around benchmarking for scale? I could add a test that does a larger computation. Another alternative would be to add a notebook with an example using a bigger dataset.

jeremy-brouillet replied:
Ran a test over a large number of examples (embedding size 4096, embeddings 200, top_k 100 and I'm seeing a 75x slowdown, I'll see if I can fix it up.

logan-markewich replied:
Thanks for checking! 🙏

jeremy-brouillet replied:
Brought back the similarity heap, doesn't look like it improves anything. As one needs to iterate through the heap and re-heapify, that was kind of expected. I'll look into more numpy-intensive approaches and see if that can help.

logan-markewich replied:
Very tricky problem due to the sorts on every loop I think 🤔 But tbh, a typical use-case might be something more like a top-k of 3, picking from 1,000ish embeddings. When I benchmarked in the past I just hard-coded the top k and scaled the embeddings lol

jeremy-brouillet replied:
I took a look at it and did some profiling. The similarity_fn, which makes a call to compare two embeddings, take up 95-99% of the compute time. For example, I'd tested an example with an embedding size of 50, 5000 embeddings, and similarity_top_k of 50. Overall, MMR is a slower algorithm than just simple similarity to the query vector, it adds additional compute steps. There could be ways to optimize past it by cutting down on the number of embedding-embedding calculations, but that goes beyond the scope of iterative MMR described in the '98 paper.

I did clean up some of the logic and remove the extra sorts. Rather than maintaining a list of results each loop, it just tracks the best result, which avoids it needing to sort. This drops the runtime down to O(n^2).

jeremy-brouillet replied:
It's just 5~10 seconds for top-100 for 1000s of embeddings of large dimensionality. So it's slower but manageable for the use cases you described.

logan-markewich replied:
Thanks for taking the time to document and profile this, it helps tons! I think this should be in a good spot to merge now 💪💪

logan-markewich commented:
One more thing, can you merge main and update the changlog.md file? Otherwise I will do that tomorrow morning lol

jeremy-brouillet commented:
I updated the changelog.md. I'm not sure that I have the write access to merge main.

jerryjliu reviewed (commented):
thanks @jeremy-brouillet! 

minor nit about notebooks: why is SimpleIndexDemo.ipynb checked in? 

also might be good to add MMR notebook here: https://gpt-index.readthedocs.io/en/latest/how_to/integrations/vector_stores.html

logan-markewich commented:
@jerryjliu I actually added a link to the notebook on that page this morning 😅

jeremy-brouillet commented:
Sure, I can move the notebook over to the more appropriate folder.

jeremy-brouillet commented:
I moved the notebook over to the vector store examples folder, added some details in the notebook, and cleaned up an edge case where the embeddings were None.

jerryjliu reviewed (approved):
thanks!
